File: apps/agents/src/open-canvas/nodes/reflect.ts
import { Client } from "@langchain/langgraph-sdk";
import { OpenCanvasGraphAnnotation } from "../state.js";
import { LangGraphRunnableConfig } from "@langchain/langgraph";

export const reflectNode = async (
  state: typeof OpenCanvasGraphAnnotation.State,
  config: LangGraphRunnableConfig
) => {
  try {
    const langGraphClient = new Client({
      apiUrl: `http://localhost:${process.env.PORT}`,
    });

    const reflectionInput = {
      messages: state._messages,
      artifact: state.artifact,
    };
    const reflectionConfig = {
      configurable: {
        // Ensure we pass in the current graph's assistant ID as this is
        // how we fetch & store the memories.
        open_canvas_assistant_id: config.configurable?.assistant_id,
      },
    };

    const newThread = await langGraphClient.threads.create();
    // Create a new reflection run, but do not `wait` for it to finish.
    // Intended to be a background run.
    await langGraphClient.runs.create(
      // We enqueue the memory formation process on the same thread.
      // This means that IF this thread doesn't receive more messages before `afterSeconds`,
      // it will read from the shared state and extract memories for us.
      // If a new request comes in for this thread before the scheduled run is executed,
      // that run will be canceled, and a **new** one will be scheduled once
      // this node is executed again.
      newThread.thread_id,
      // Pass the name of the graph to run.
      "reflection",
      {
        input: reflectionInput,
        config: reflectionConfig,
        // This memory-formation run will be enqueued and run later
        // If a new run comes in before it is scheduled, it will be cancelled,
        // then when this node is executed again, a *new* run will be scheduled
        multitaskStrategy: "enqueue",
        // This lets us "debounce" repeated requests to the memory graph
        // if the user is actively engaging in a conversation. This saves us $$ and
        // can help reduce the occurrence of duplicate memories.
        afterSeconds: 5 * 60, // 5 minutes
      }
    );
  } catch (e) {
    console.error("Failed to start reflection");
  }

  return {};
};

File: agents-py/agents/src/open_canvas/nodes/reflect.py
import os
from typing import Dict, Any
from langgraph_sdk import get_client
from langchain_core.runnables import RunnableConfig
from agents.src.open_canvas.state import OpenCanvasGraphState

async def reflect(
    state: OpenCanvasGraphState,
    config: RunnableConfig
) -> Dict[str, Any]:
    try:
        lang_graph_client = get_client(url=f"http://localhost:{os.getenv('PORT', '8000')}")

        reflection_input = {
            "messages": state._messages,
            "artifact": state.artifact.dict() if state.artifact else None
        }
        
        reflection_config = {
            "configurable": {
                "open_canvas_assistant_id": config.get("configurable", {}).get("assistant_id")
            }
        }

        new_thread = await lang_graph_client.threads.create()
        
        await lang_graph_client.runs.create(
            thread_id=new_thread.thread_id,
            app_id="reflection",
            input=reflection_input,
            config=reflection_config,
            multitask_strategy="enqueue",
            after_seconds=5 * 60  # 5 minutes
        )
        
    except Exception as e:
        print(f"Failed to start reflection\n{e}")

    return {} 